src/operator/tensor/control_flow_op.h

```diff
@@ -0,0 +1,226 @@
+ * The condition, x, and y have the same shape.
+ * The returned array is formed by elements from x or y
+ * depending on the elements of condition.
+ */
+template<int req>
+struct where {
+  // DType is the output data type
+  // CType is condition data type
+  template<typename DType, typename CType>
+  MSHADOW_XINLINE static void Map(int i, DType* out, const CType* cond,
+                                  const DType* x, const DType* y) {
+    KERNEL_ASSIGN(out[i], req, (0 != cond[i]? x[i] : y[i]));
+  }
+};
+
+/*! \brief Choose elements from x or y depending on condition
+ * The condition is a vector whose size is the same as the
+ * x's first dim size.
+ * The returned array is formed by rows from x or y depending on
+ * the condition's elements.
+ */
+template<int req>
+struct where_batch {
+  // DType is the output data type
+  // CType is the condition data type
+  template<typename DType, typename CType>
+  MSHADOW_XINLINE static void Map(int i, DType* out, const CType* cond,
+                                  const DType* x, const DType* y, int M) {
+    KERNEL_ASSIGN(out[i], req, (0 != cond[i/M]? x[i] : y[i]));
+  }
+};
+
+/*!
+ * \brief Template for calculating grad[x] and grad[y].
+ * template argument req is OpReqType; negate indicates
+ * whether the output is grad_x (negate=true)
+ * or grad_y (negate=false).
+ */
+template<int req, bool negate>
+struct where_backward {
+  // DType is the output data type
+  // CType is condition data type
+  template<typename DType, typename CType>
+  MSHADOW_XINLINE static void Map(int i, DType* grad_out,
+                                  const DType* grad_in,
+                                  const CType* cond) {
+    KERNEL_ASSIGN(grad_out[i], req,
+      ((0 == cond[i])^negate)? grad_in[i] : static_cast<DType>(0));
+  }
+};
+
+/*!
+ * \brief Template for calculating grad[x] and grad[y].
+ * template argument req is OpReqType; negate indicates
+ * whether the output is grad_x (negate=true)
+ * or grad_y (negate=false).
+ * The condition is a vector whose size is the same as the
+ * x's first dim size.
+ */
+template<int req, bool negate>
+struct where_batch_backward {
+  // DType is the output data type
+  // CType is condition data type
+  template<typename DType, typename CType>
+  MSHADOW_XINLINE static void Map(int i, DType* grad_out,
+                                  const DType* grad_in,
+                                  const CType* cond, int M) {
+    KERNEL_ASSIGN(grad_out[i], req,
+      ((0 == cond[i/M])^negate)? grad_in[i] : static_cast<DType>(0));
+  }
+};
+
+inline bool WhereOpShape(const nnvm::NodeAttrs& attrs,
+                         std::vector<TShape>* in_attrs,
+                         std::vector<TShape>* out_attrs) {
+  CHECK_EQ(in_attrs->size(), 3)
+    << "where operator takes 3 arguments (" << in_attrs->size() << " given)";
+  CHECK_EQ(out_attrs->size(), 1);
+
+  TShape tshape((*in_attrs)[1]);
+  if (!shape_assign(&tshape, (*in_attrs)[2])) return false;
```

Comment:
return false means missing information. If shape_assign fails it means shapes are incompatible and should fail